sopel: drop support for Python 3.6 and require 3.7+

Exirel opened the pull request:
### Description

Tin. Python 3.6 died yesterday. I checked the warning for 3.6, they were written for Sopel 8, so we can either keep 3.6 or drop it now. Either way, this is not a big change, unlike Python 2.

I haven't modified the usage of `OrderedDict`, because there are slight difference between `dict` and `OrderedDict` in particular when considering equality. 

I haven't check the "future" import either, hence the "draft" status for now.

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

dgw commented:
Might as well take this opportunity to add 3.10 to CI, no?

Exirel commented:
>  Test suite (3.6) Expected — Waiting for status to be reported 

I've no idea what I missed here.

dgw commented:
> >  Test suite (3.6) Expected — Waiting for status to be reported 
> 
> I've no idea what I missed here.

It's because that check is marked as required in the repo settings. Before merging I would just have to update the list of expected checks.

dgw commented on .github/workflows/ci.yml:
```diff
@@ -20,7 +20,6 @@ jobs:
-          - 3.6
           - 3.7
           - 3.8
           - 3.9
```
```suggestion
          - 3.9
          - 3.10
```

dgw commented on setup.cfg:
```diff
@@ -16,14 +16,13 @@ classifiers =
-    Programming Language :: Python :: 3.6
     Programming Language :: Python :: 3.7
     Programming Language :: Python :: 3.8
     Programming Language :: Python :: 3.9
```
```suggestion
    Programming Language :: Python :: 3.9
    Programming Language :: Python :: 3.10
```

dgw commented on sopel/bot.py:
```diff
@@ -1157,8 +1157,8 @@ def search_url_callbacks(self, url):
             The Python documentation for the `re.search`__ function and
             the `match object`__.
 
-        .. __: https://docs.python.org/3.6/library/re.html#re.search
-        .. __: https://docs.python.org/3.6/library/re.html#match-objects
+        .. __: https://docs.python.org/3.7/library/re.html#re.search
+        .. __: https://docs.python.org/3.7/library/re.html#match-objects
```
Surely at least `re.search` could use an intersphinx reference instead of a ha(n|r)d-coded URL? It should be possible to <tt>:ref:\`match-objects\`</tt>, too (also in `sopel/plugins/rules.py`).

Perhaps should be changed in one of the doc cleanups you/we are planning, instead of this PR, but I'm still going to call attention to it because it's being touched here. 😁

Exirel replied:
Probably. I'll leave that to a future PR, one that fix all the "Pattern" and "Match" not found errors. 😁

Exirel replied:
Future PR then?

dgw replied:
Yeah, future is fine.

dgw reviewed (commented):
Suggestions from me should be the only changes needed to add Python 3.10 to CI and reflect its "officially supported" status in the package once built.

Exirel commented:
So Github CI thinks it's "3.1" and not "3.10"... weird.

Exirel commented on .github/workflows/ci.yml:
```diff
@@ -20,10 +20,10 @@ jobs:
           - 3.7
           - 3.8
           - 3.9
+          - 3.10
```
```suggestion
          - '3.10'
```

Let's try with a quote, because apparently YAML thinks it's a float... And people wonder why I hate YAML...

Exirel commented:
Alright, now we can see that Python 3.10 is not supported by `requests` yet. See also: https://pyreadiness.org/3.10/

dgw commented:
I see a lot of transitive requirements in the "not supported" category, too—stuff we don't use directly, but that things we require depend on.

Let's abandon adding 3.10 in this patch and try again after the next `requests` release. (2.27 is supposed to be the last that will support Python 2.7, too, so that will be time to revisit our 7.x branch.)

Exirel commented:
Weird, there is an isup test that fails in all three versions. I'm looking at potential cause, so here is the version installed:

- requests-2.27.1
- requests-mock~=1.9.3

I'll test locally to see if I can reproduce or not.

Exirel commented:
I can confirm that with `request` 2.26 the test pass, not with 2.27. This will need another PR to fix that.

Exirel commented:
Following [Python 3.7 documentation](https://docs.python.org/3.7/library/__future__.html), I removed the `generator_stop` and added the `annotation` future imports.

dgw commented on setup.cfg:
```diff
@@ -50,12 +49,11 @@ ignore =
-    FI10,FI11,FI12,FI13,FI14,FI16,FI17,
+    # oldest supported Python version (3.7).
+    FI10,FI11,FI12,FI13,FI14,FI15,FI16,FI17,
     # Ignore "annotations" future import, since it's not available before 3.7
```
```suggestion
```

We're not ignoring this one any more. Don't know if trailing comma is allowed in this value, so you might need to remove the comma after `FI17` above.

dgw reviewed (commented):
One nit, for your INI-tweaking pleasure.

Exirel commented:
Yup, fixed it and squashed it. Didn't change anything else.

dgw reviewed (approved):
I guess passing CI is our proof that trailing commas in flake8 config don't matter. 😁
